本次分享主要内容是大模型最核心的地方：自注意力机制。
![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image001.png)

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image003.png)

接下来会分四部分，第一部分是自注意力机制简化版，第二部分是自注意力机制，第三部分是因果注意力(causal attention)，第四部分是多头注意力(multi-head attention)。

#### 1.The problem with modeling long sequences
让咱们看下在没有自注意力机制之前如何解决长序列问题的，这里咱们拿语言翻译举例。

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image005.png)

上图是德文翻译成英文的例子，发现单词不能逐词翻译，不然就会乱套，因为每种语言的语法不一样的，这需要模型能够看到上下文，才能做出正确的翻译。</br>

那么自然想到的办法是使用encoder-decoder架构，先使用encoder看到所有输入，理解上下文，然后使用decoder逐词输出。</br>

在transformer架构出来之前，最流行的encoder-decoder架构是RNN，比如LSTM和GRU。


![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image007.png)

如上图所示，所有的输入先经过RNN encoder，得到一个隐状态，在encoder里，输出是下一步的输入，所以隐状态是包含了上下文的。然后隐状态作为decoder的输入。

可以简单把hidden state理解为包含了上下文的embedding vector。

RNN的缺点是，decoder无法获取之前的隐状态信息，所以需要把所有的隐状态都保存下来，这样会占用大量内存。同时长距离文本处理起来比较困难。



#### 2.Capturing data dependencies with self-attention

2014年，研究人员提出了一个Bahdanau Attention的机制，通过修改encoder-decoder架构，以便decoder能够有选择的获得输入序列的不同部分。

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image009.png)

如上图，在decoder阶段，针对特定输出，可以获得输入序列的不同部分，并通过attention weight来衡量输入序列不同部分对当前输出的重要性。

Bahdanau Attention机制的缺点是，无法并行处理，在encoder阶段，所有的输入必须一个一个处理，在decoder阶段，当前时刻的隐状态依赖于前一个时刻的隐状态，所以也无法并行处理。

三年后2017年，受到了Bahdanau Attention的启发，提出了transformer架构，使用自注意力机制来解决长距离依赖问题。

Self-attention（自注意力）是一种让输入序列中的每个位置都能够在计算序列表示时关注同一序列中所有位置的机制。这种机制使得模型能够捕捉序列内部的长距离依赖关系，并且理解不同位置之间的复杂关系。

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image011.png)

#### 3.Attending to different parts of the input with self-attention

本章会介绍自注意力机制内部工作机制，以及学习如何从头实现自注意力机制。自注意力机制作为大模型的基石，非常值得花费时间去理解，一旦掌握了机制背后的原理，对后续的深入学习非常有帮助。

self-attention里的self表示什么意思呢？

“Self”在自注意力中指的是机制能够计算不同位置之间的注意力权重，这些位置都位于单个输入序列内部。这意味着模型在处理序列时，会分析序列内部各部分之间的关系，例如句子中的单词或图像中的像素。

自注意力机制评估并学习输入序列内部各部分之间的关系和依赖性。这种机制使得模型能够理解序列中各个元素之间的相互作用，而不仅仅是序列作为一个整体与其他序列之间的关系。

传统的注意力机制通常关注两个不同序列之间的元素关系，例如在序列到序列（sequence-to-sequence）模型中，注意力可能集中在输入序列和输出序列之间的关系。这种注意力机制在机器翻译等任务中很常见，其中模型需要关注输入序列（如一种语言的文本）与输出序列（如另一种语言的文本）之间的关系，比如上图所示。

自注意力在自然语言处理（NLP）和计算机视觉（CV）等领域有广泛应用。在NLP中，它可以用于理解文本中单词之间的关系；在CV中，它可以用于理解图像中不同像素或区域之间的关系。

自注意力机制比较复杂，为了让大家更方便的理解其原理，我们先实现一个简单的没有可训练权重的自注意力机制，然后逐步增加可训练权重，最后实现一个完整的自注意力机制。


##### 3.1 A simple self-attention mechanism without trainable weights

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image013.png)

自注意力机制的目标是为输入序列中的每个元素计算一个上下文向量(context vector),这个向量结合了来自序列中所有其他元素的信息。假如咱们要对第二个元素${x^{(2)}}$计算上下文向量${z^{(2)}}$,每个输入元素对于计算上下文向量${z_2}$的重要性或者贡献度由注意力权重${a_{21}}$到${a_{2T}}$决定。这里的T表示输入序列的长度，而${a_{2i}}$表示第i个元素对第2个元素的注意力权重。上下文向量${z^{(2)}}$的计算公式如下：

$$z^{(2)} = \sum_{i=1}^{T} a_{2i} x^{(i)}$$

为了计算注意力权重，咱们需要一个打分函数，这个函数的作用是衡量两个元素之间的相似度。常用的打分函数是点积，即两个向量对应元素相乘然后求和。

$$e_{ij} = x_i^T x_j$$

In [72]:
import torch
inputs = torch.tensor(
  [[0.43, 0.15, 0.89], # Your     (x^1)
   [0.55, 0.87, 0.66], # journey  (x^2)
   [0.57, 0.85, 0.64], # starts   (x^3)
   [0.22, 0.58, 0.33], # with     (x^4)
   [0.77, 0.25, 0.10], # one      (x^5)
   [0.05, 0.80, 0.55]] # step     (x^6)
)

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image015.png)

这里我们将输入元素${x^{(2)}}$叫作query,为什么叫作query，留到下回分解。

我们先来看下中间一步的计算，计算attention score, 即query ${x^{(2)}}$ 与其他输入元素的点积。

点积计算公式如下：
$$e_{ij} = x_i^T x_j$$

也就是两个向量的元素对应相乘然后求和。

In [73]:
query = inputs[1]
atten_scores_2 = torch.empty(inputs.shape[0])
for i in range(inputs.shape[0]):
    atten_scores_2[i] = torch.dot(query, inputs[i])
print(atten_scores_2)

tensor([0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865])


In [74]:
res = 0
for idx, element in enumerate(inputs[0]):
    res += element * query[idx]
print(res)
print(torch.dot(inputs[0], query))

tensor(0.9544)
tensor(0.9544)


点积其实是衡量两个向量相似度的方法之一，点积越大，两个向量越相似。这里使用到点积来衡量一个文本序列里，不同token之间的关系，点积越大，关系越密切。
![](./image.png)

下一步就是通过归一化attention score，得到注意力权重attention weight。

![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image017.png)

归一化公式如下：
$$a_{ij} = \frac{{w_{ij}}}{\sum_{k=1}^{T} w_{ik}}$$

归一化的目的是让注意力权重之和为1，这样就可以保证所有注意力权重加起来等于1。归一化对可解释性和训练稳定性都有好处。

In [75]:
atten_weights_2_tmp = atten_scores_2 / atten_scores_2.sum()
print("Attention weights:",atten_weights_2_tmp)
print("Sum:",atten_weights_2_tmp.sum())

Attention weights: tensor([0.1455, 0.2278, 0.2249, 0.1285, 0.1077, 0.1656])
Sum: tensor(1.0000)


在实际工程中，使用softmax函数来做归一化更为常见，这种方法对于极大或者极小的值更为鲁棒，同时还能提供更有利的梯度特性。softmax函数公式如下：

$$a_{ij} = \frac{{e^{w_{ij}}}}{\sum_{k=1}^{T} e^{w_{ik}}}$$

In [76]:
def softmax_naive(x):
    return torch.exp(x) / torch.exp(x).sum(dim=0)
atten_weights_2_naive = softmax_naive(atten_scores_2)
print("Attention weights:",atten_weights_2_naive)
print("Sum:",atten_weights_2_naive.sum())

Attention weights: tensor([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
Sum: tensor(1.)


softmax方法可以确保attention weight总是正的，而除以sum的方法无法保证这一点。这点可以让attention weight可以看作概率分布或者相对重要性。

softmax_naive方法的缺点是，对于极大或者极小的值，计算结果不稳定，容易出现数值溢出。因此建议使用PyTorch自带的softmax函数。

In [77]:
atten_weights_2 = torch.softmax(atten_scores_2, dim=0)
print("Attention weights:",atten_weights_2)
print("Sum:",atten_weights_2.sum())

Attention weights: tensor([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
Sum: tensor(1.)


![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image019.png)

最后一步就是计算上下文向量context vector。

计算公式如下：

$$z^{(2)} = \sum_{i=1}^{T} a_{2i} x^{(i)}$$

每一个输入向量乘以对应的attention weight，然后求和，得到上下文向量。

In [78]:
query = inputs[1]
context_vector_2 = torch.zeros(query.shape)
for i, x_i in enumerate(inputs):
    context_vector_2 += atten_weights_2[i] * x_i
print("Context vector:",context_vector_2)

Context vector: tensor([0.4419, 0.6515, 0.5683])


#### 3.2 Compuing attention weights for all input tokens

为了计算所有输入元素的上下文向量，咱们需要对每个输入元素重复上述步骤。

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-11.png)

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-12.png)

我们重复上述步骤，得到所有输入元素的上下文向量。


In [79]:
atten_scores = torch.empty(inputs.shape[0], inputs.shape[0])
for i, x_i in enumerate(inputs):
    for j, x_j in enumerate(inputs):
        atten_scores[i, j] = torch.dot(x_i, x_j)
print(atten_scores)

tensor([[0.9995, 0.9544, 0.9422, 0.4753, 0.4576, 0.6310],
        [0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865],
        [0.9422, 1.4754, 1.4570, 0.8296, 0.7154, 1.0605],
        [0.4753, 0.8434, 0.8296, 0.4937, 0.3474, 0.6565],
        [0.4576, 0.7070, 0.7154, 0.3474, 0.6654, 0.2935],
        [0.6310, 1.0865, 1.0605, 0.6565, 0.2935, 0.9450]])


In [80]:
# 上面是通过for循环来进行计算的，性能其实很低下，我们可以使用矩阵乘法来优化
# inputs是(6, 3)的矩阵，inputs.T是(3, 6)的矩阵，所以atten_scores是(6, 6)的矩阵
atten_scores = inputs @ inputs.T
print(atten_scores)

tensor([[0.9995, 0.9544, 0.9422, 0.4753, 0.4576, 0.6310],
        [0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865],
        [0.9422, 1.4754, 1.4570, 0.8296, 0.7154, 1.0605],
        [0.4753, 0.8434, 0.8296, 0.4937, 0.3474, 0.6565],
        [0.4576, 0.7070, 0.7154, 0.3474, 0.6654, 0.2935],
        [0.6310, 1.0865, 1.0605, 0.6565, 0.2935, 0.9450]])


In [81]:
#接下来咱们把atten_scores归一化，获得attention weight
#dim=-1表示在最后一个维度上进行归一化（这里我就写一个了“#”，cursor就知道我要说明dim的含义）
atten_weights = torch.softmax(atten_scores, dim=-1)
print(atten_weights)


tensor([[0.2098, 0.2006, 0.1981, 0.1242, 0.1220, 0.1452],
        [0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581],
        [0.1390, 0.2369, 0.2326, 0.1242, 0.1108, 0.1565],
        [0.1435, 0.2074, 0.2046, 0.1462, 0.1263, 0.1720],
        [0.1526, 0.1958, 0.1975, 0.1367, 0.1879, 0.1295],
        [0.1385, 0.2184, 0.2128, 0.1420, 0.0988, 0.1896]])


In [82]:
# 验证下每一行和是否为1
row_2_sum = sum([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
print(row_2_sum)
print("All row sums:", atten_weights.sum(dim=-1))

1.0
All row sums: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000])


In [83]:
# 最后一步，通过矩阵乘法计算context vector
# atten_weights是(6, 6)的矩阵，inputs是(6, 3)的矩阵，所以all_context_vectors是(6, 3)的矩阵
all_context_vectors = atten_weights @ inputs
print(all_context_vectors)

tensor([[0.4421, 0.5931, 0.5790],
        [0.4419, 0.6515, 0.5683],
        [0.4431, 0.6496, 0.5671],
        [0.4304, 0.6298, 0.5510],
        [0.4671, 0.5910, 0.5266],
        [0.4177, 0.6503, 0.5645]])


In [84]:
# 看看上节课计算的context vector
print("Previous 2nd context vector:",context_vector_2)

Previous 2nd context vector: tensor([0.4419, 0.6515, 0.5683])


接下来咱们开始增加可训练权重，看看如何实现一个完整的自注意力机制。

#### 4.Implementing self-attention with trainable weights

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-13.png)

接下来真正开启自注意力机制的实现，这也是GPT-2和GPT-3的核心，也是其他类似大模型的核心机制，通过上面的学习，可以知道自注意力机制也叫做缩放点积注意力机制(scaled dot-product attention)。

带可训练权重的自注意力机制和之前相比，最重要的区别是权重矩阵在训练过程中是可以变化的。 这些权重矩阵之所以重要是因为他们可以产生“好的”的context vectors。

接下来咱们通过两个小节来介绍，第一个小节和之前一样，一步步的来把自注意力机制的各个部分实现出来，第二小节会把把这些代码封装成一个python类，以便可以用于大模型的框架中。

#### 4.1一步步计算注意力权重

这里首先思考下什么是可训练权重？
我们先介绍三个可训练权重矩阵，分别是query矩阵($W_q$)，key矩阵($W_k$)和value矩阵($W_v$)。这三个矩阵被用来将嵌入式输入tokensz转为query,key和value向量。如下图所示：
![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-14.png)

类似上节课所讲的，我们将$x^{(2)}$作为query input,query vector为$q^{(2)}$，计算公式如下：
$$q^{(2)} = x^{(2)} W_q$$

同理，key vector和value vector的计算公式如下：
$$k^{(2)} = x^{(2)} W_k$$
$$v^{(2)} = x^{(2)} W_v$$

就像上节课所讲，先计算$x^{(2)}$对应的context vector,然后再计算所有输入元素的context vector。

In [85]:
# 先定义一些变量
x_2 = inputs[1]
d_in = inputs.shape[1]
d_out = 2

这里，输出变量的维度d_out和输入变量的维度d_in可以不同，也可以相同。GPT的输入是token embedding，维度是d_in，输出的context vector维度也是d_in。

In [86]:
# 初始化query,key和value权重矩阵
torch.manual_seed(123)
W_query = torch.nn.Parameter(torch.rand(d_in, d_out),requires_grad=False)
W_key = torch.nn.Parameter(torch.rand(d_in, d_out),requires_grad=False)
W_value = torch.nn.Parameter(torch.rand(d_in, d_out),requires_grad=False)

torch.nn.Parameter是PyTorch中的一个类，用于创建可训练的参数。torch.nn.Parameter是torch.Tensor的子类，它与torch.Tensor的区别在于，torch.nn.Parameter的默认requires_grad属性为True，而torch.Tensor的默认requires_grad属性为False。这意味着，当使用torch.nn.Parameter创建一个张量时，PyTorch会自动将其标记为需要梯度，以便在反向传播过程中进行优化。
 
这里设置requires_grad=False，表示这些权重矩阵在训练过程中不会被更新。是为了减小计算量，但在模型训练中，我们需要设置requires_grad=True，以便可以更新这些权重矩阵。

In [87]:
#计算query,key和value向量
query_2 = x_2 @ W_query
key_2 = x_2 @ W_key
value_2 = x_2 @ W_value
print("input:",x_2)
print("Query:",query_2)
print("Key:",key_2)
print("Value:",value_2)

input: tensor([0.5500, 0.8700, 0.6600])
Query: tensor([0.4306, 1.4551])
Key: tensor([0.4433, 1.1419])
Value: tensor([0.3951, 1.0037])


这里再说明下权重参数和上节课讲的attention weight的区别，权重参数是可训练的，而attention weight是可训练权重经过softmax归一化后的结果,attention weight决定了每个输入元素对当前元素的贡献度。

In [88]:
#计算所有输入元素对应的keys和values
keys = inputs @ W_key
values = inputs @ W_value
print("keys.shape:",keys.shape)
print("values.shape:",values.shape)
print(keys)

keys.shape: torch.Size([6, 2])
values.shape: torch.Size([6, 2])
tensor([[0.3669, 0.7646],
        [0.4433, 1.1419],
        [0.4361, 1.1156],
        [0.2408, 0.6706],
        [0.1827, 0.3292],
        [0.3275, 0.9642]])


可以看到我们成功的把输入的embedding从3维度转换到了2维度。

下一步就是计算attention score，即query和key的点积。如下图所示：

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-15.png)
![](https://drek4537l1klr.cloudfront.net/raschka/v-8/Figures/ch03__image015.png)

和简化版自注意力机制不同一点是，attention score不是通过输入元素间的点积获得的，而是通过query和key的点积获得的。


首先先计算attention score $w_{22}$

In [89]:
keys_2 = keys[1]
attn_score_22 = query_2.dot(keys_2)
print(attn_score_22)

tensor(1.8524)


In [90]:
#计算所有的attention_scores通过矩阵相乘
attn_scores_2 = query_2 @ keys.T
print(attn_scores_2)

tensor([1.2705, 1.8524, 1.8111, 1.0795, 0.5577, 1.5440])


可以看到，attn_score_2第二个元素和attn_score_22一样的，都是1.8524

下一步就是从attention scores获得attention weights,上节课咱们直接使用pytorch提供的softmax方法来将attention scores转换为attention weights，但这里我们新增了一个环节，就是在使用softmax方法之前，先对attention scores进行缩放，也就是attention scores除以输入向量维度的平方根。

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-16.png)


In [91]:
d_k = keys.shape[-1]
attn_weights_2 = torch.softmax(attn_scores_2 / d_k**0.5,dim=-1)
print(attn_weights_2)

tensor([0.1500, 0.2264, 0.2199, 0.1311, 0.0906, 0.1820])


![](https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSzJosiqrYDtp_Tv6cQWh8hiqORrOZcjTF-jbKefyLUZlWCRadPnxgNQOdu_nfE9Fzql30&usqp=CAU)

为什么这么做呢？大家可以看下softmax函数的图形，发现x自变量的值越大，softmax的值变化越小，也就是梯度越小，这时候就会带来梯度消失的问题，导致无法训练成功后。

当嵌入维度（embedding dimension）很大时（例如GPT等大型语言模型通常使用的维度大于1000），点积（dot products）可能会变得非常大。这会导致在反向传播过程中，经过softmax函数处理后的梯度变得非常小。

为了解决这个问题，自注意力机制中的点积被缩放，通常是除以嵌入维度的平方根。这样做可以防止点积过大，从而避免softmax函数输出的梯度过小。

由于这种缩放操作，自注意力机制也被称为缩放点积注意力（scaled-dot product attention）。这种缩放有助于保持梯度在一个合理的范围内，从而有助于模型更有效地学习。

通过这种缩放，可以改善训练性能，使得模型在训练过程中能够更稳定地收敛，并且可以更快地学习到数据中的复杂模式。

至于为什么选用嵌入维度的平方根呢？这个缩放因子是基于实验和经验得出的。在实践中，它被证明能够提高模型的训练效率和性能。从信息论的角度来看，缩放点积可以帮助保持信息的方差，从而有助于模型学习到更丰富的表示。理论上，可以使用其他缩放因子，例如嵌入维度本身或其他常数。但是，嵌入维度的平方根被广泛认为是一个好的选择，因为它在实践中表现良好。

这也就是为什么大模型很多时候靠实践的，是不是有点像中医的理论？

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-17.png)

下一步就是计算context vectors

和上节课类似，context vectors是attention weights作为权重，乘以values vector,而不是输入向量了。

注意力权重决定了每个值向量在计算上下文向量时的相对重要性。这些权重是通过模型学习得到的，它们反映了模型对序列中不同部分的关注程度。


In [92]:
context_vec_2 = attn_weights_2 @ values
print(context_vec_2)

tensor([0.3061, 0.8210])


#### 4.2 Implementing a compact self-attention Python class

上面为了给大家演示原理，是逐行解释代码的，但在实际工作中，需要将上面的代码封装进python类中，这个小节就是展示如何编写一个自注意力机制的python类。

In [93]:
import torch.nn as nn
class SelfAttention_v1(nn.Module):
    def __init__(self,d_in,d_out):
        super().__init__()
        self.W_query = nn.Parameter(torch.rand(d_in,d_out))
        self.W_key = nn.Parameter(torch.rand(d_in,d_out))
        self.W_value = nn.Parameter(torch.rand(d_in,d_out))
        
    
    def forward(self,x):
        keys = x @ self.W_key
        queries = x @ self.W_query
        values = x @ self.W_value
        attn_scores = queries @ keys.T

        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5,dim=-1
        )
        context_vec = attn_weights @ values
        return context_vec

In [159]:
torch.manual_seed(123)
sa_v1 = SelfAttention_v1(d_in,d_out)
print(sa_v1(inputs))

tensor([[0.2996, 0.8053],
        [0.3061, 0.8210],
        [0.3058, 0.8203],
        [0.2948, 0.7939],
        [0.2927, 0.7891],
        [0.2990, 0.8040]], grad_fn=<MmBackward0>)


在实际项目中，我们一般使用nn.Linear层来替换nn.Parameter方法。nn.Linear实现的数学表达为 y = wx + b.如果b没有的话，这时候nn.Linear实现的功能就是和nn.Parameter一样，同时nn.Linear可以自动实现矩阵相乘并且在初始化权重参数时候更优，能够让训练更稳定和有效。

In [95]:
class SelfAttention_v2(nn.Module):
    def __init__(self,d_in,d_out,qkv_bias=False):
        super().__init__()
        self.W_query = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.W_key = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.W_value = nn.Linear(d_in,d_out,bias=qkv_bias)
    
    def forward(self, x):
        keys = self.W_key(x)
        queries = self.W_query(x)
        values = self.W_value(x)
        attn_scores = queries @ keys.T
        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5,dim=-1
        )
        context_vec = attn_weights @ values
        return context_vec

In [101]:
torch.manual_seed(123)
sa_v2 = SelfAttention_v2(d_in,d_out)
print(sa_v2(inputs))

tensor([[-0.5337, -0.1051],
        [-0.5323, -0.1080],
        [-0.5323, -0.1079],
        [-0.5297, -0.1076],
        [-0.5311, -0.1066],
        [-0.5299, -0.1081]], grad_fn=<MmBackward0>)


可以看到SelfAttention_v1和SelfAttention_v2输出结果不一样，这是因为nn.Linear使用了更复杂的权重初始化机制。

In [97]:
# 练习SelfAttention_v1和SelfAttention_v2互换
import torch.nn as nn
class SelfAttention_v3(nn.Module):
    def __init__(self,P_query,P_key,P_value):
        super().__init__()
        self.W_query = P_query
        self.W_key = P_key
        self.W_value = P_value
        
    def forward(self,x):
        keys = x @ self.W_key
        queries = x @ self.W_query
        values = x @ self.W_value
        attn_scores = queries @ keys.T

        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5,dim=-1
        )
        context_vec = attn_weights @ values
        return context_vec

In [100]:
torch.manual_seed(123)
sa_v2 = SelfAttention_v2(d_in,d_out)
P_query = sa_v2.W_query.weight.T
P_key = sa_v2.W_key.weight.T
P_value = sa_v2.W_value.weight.T
sp_v3 = SelfAttention_v3(P_query,P_key,P_value)
print(sp_v3(inputs))

tensor([[-0.5337, -0.1051],
        [-0.5323, -0.1080],
        [-0.5323, -0.1079],
        [-0.5297, -0.1076],
        [-0.5311, -0.1066],
        [-0.5299, -0.1081]], grad_fn=<MmBackward0>)


接下来会重点讲解causal和multi-head elements。其中casual机制阻止模型学习到未来的信息，因为预测下一个token只能依赖之前的token，没有人能够预知未来。
multi-head是将一个注意力模型分为多个不同的模型，注意力机制（Attention Mechanism）是一种用于聚焦输入序列不同部分的重要技术。多头组件将这个注意力机制拆分成多个 “头（heads）”。例如，假设我们有一个 8 头的多头注意力机制，就相当于把原本单一的注意力计算过程复制成了 8 个独立的部分。
每个头的学习任务：每个头会学习数据的不同方面。以处理一段文本为例，一个头可能会更关注文本中的语法结构相关的信息，比如句子的主谓宾成分；另一个头可能会侧重于词汇的语义信息，像词语的情感色彩。这些头是并行工作的，它们在处理相同的输入序列时，会根据自己学到的侧重点来提取和关注信息。

#### 5 hiding future words with causal attention
如上面所说，我们不希望预测下一个token时候，使用到未来的信息，所以我们需要修改上面的自注意力机制，只注意到之前的信息，这种方式被称为因果注意力机制，也被称为掩码注意力机制(masked attention).

在 Transformer 架构的解码器（Decoder）部分广泛使用了 masked attention。在生成序列（如文本生成）的过程中，模型在预测下一个单词时，不应该看到未来的单词信息。例如，当生成一个句子 “The dog is running” 时，在预测 “is” 这个词的时候，不能让模型 “看到”“running” 这个词。

这时候就需要使用 masked attention。通过创建一个掩码（mask）来遮蔽（或掩盖）未来的信息。这个掩码通常是一个上三角矩阵（在自注意力机制的情况下），在矩阵中，对应未来位置的元素被设置为 0 或者负无穷（取决于具体实现），这样在计算注意力权重时，这些被遮蔽的未来位置的信息就不会被考虑进去。

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-19.png)

在因果注意力中，我们将对角线上方的注意力权重屏蔽掉，这样对于给定的输入，语言模型在使用注意力权重计算上下文向量时无法访问未来的标记。例如，对于第二行中的单词 “journey”，我们只保留其前面（“Your”）和当前位置（“journey”）的注意力权重。

同时还需要归一化非被掩码的注意力权重，以让这些权重和为1在每一行。

##### 5.1 Applying a causal attention mask

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-20.png)

在因果注意力中，获得掩码注意力权重矩阵的一种方法是将注意力得分应用 softmax 函数，将对角线上方的元素置零，然后对得到的矩阵进行归一化.

问题：为什么先用softmax函数呢？

In [102]:
# 第一步先和以前一样计算出了权重
queries = sa_v2.W_query(inputs)
keys = sa_v2.W_key(inputs)
attn_scores = queries @ keys.T
attn_weights = torch.softmax(attn_scores/keys.shape[-1]**0.5,dim=-1)
print(attn_weights)

tensor([[0.1717, 0.1762, 0.1761, 0.1555, 0.1627, 0.1579],
        [0.1636, 0.1749, 0.1746, 0.1612, 0.1605, 0.1652],
        [0.1637, 0.1749, 0.1746, 0.1611, 0.1606, 0.1651],
        [0.1636, 0.1704, 0.1702, 0.1652, 0.1632, 0.1674],
        [0.1667, 0.1722, 0.1721, 0.1618, 0.1633, 0.1639],
        [0.1624, 0.1709, 0.1706, 0.1654, 0.1625, 0.1682]],
       grad_fn=<SoftmaxBackward0>)


In [103]:
# 通过PyTorch的tril(下三角阵,matlab常用命令)方法可以创建一个对角线以上的值为0的mask矩阵

context_length = attn_scores.shape[0]
mask_simple = torch.tril(torch.ones(context_length,context_length))
print(mask_simple)

tensor([[1., 0., 0., 0., 0., 0.],
        [1., 1., 0., 0., 0., 0.],
        [1., 1., 1., 0., 0., 0.],
        [1., 1., 1., 1., 0., 0.],
        [1., 1., 1., 1., 1., 0.],
        [1., 1., 1., 1., 1., 1.]])


In [104]:
# 我们可以使用“*”操作让权重矩阵和掩码矩阵逐个元素相乘，以让权重矩阵上三角元素为0
masked_simple = attn_weights * mask_simple
print(masked_simple)

tensor([[0.1717, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.1636, 0.1749, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.1637, 0.1749, 0.1746, 0.0000, 0.0000, 0.0000],
        [0.1636, 0.1704, 0.1702, 0.1652, 0.0000, 0.0000],
        [0.1667, 0.1722, 0.1721, 0.1618, 0.1633, 0.0000],
        [0.1624, 0.1709, 0.1706, 0.1654, 0.1625, 0.1682]],
       grad_fn=<MulBackward0>)


In [106]:
# 最后一步是重新归一化权重矩阵
row_sums = masked_simple.sum(dim=-1,keepdim=True)
print(row_sums)
masked_simple_norm = masked_simple / row_sums
print(masked_simple_norm)

tensor([[0.1717],
        [0.3385],
        [0.5132],
        [0.6693],
        [0.8361],
        [1.0000]], grad_fn=<SumBackward1>)
tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.4833, 0.5167, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3190, 0.3408, 0.3402, 0.0000, 0.0000, 0.0000],
        [0.2445, 0.2545, 0.2542, 0.2468, 0.0000, 0.0000],
        [0.1994, 0.2060, 0.2058, 0.1935, 0.1953, 0.0000],
        [0.1624, 0.1709, 0.1706, 0.1654, 0.1625, 0.1682]],
       grad_fn=<DivBackward0>)


![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-21.png)

在因果注意力中，获得掩码注意力权重矩阵的一种更有效的方法是在应用 softmax 函数之前，用负无穷大的值来掩盖注意力得分。

在softmax方法中将负无穷大视为概率0.

softmax方法计算公式为：$$a_{ij} = \frac{{e^{w_{ij}}}}{\sum_{k=1}^{T} e^{w_{ik}}}$$

In [120]:
# 我们可以使用triu(上三角矩阵)方法,diagonal参数用于指定对角线的偏移量
mask = torch.triu(torch.ones(context_length,context_length),diagonal=1)
print(mask)
print(mask.bool())
masked = attn_scores.masked_fill(mask.bool(),-torch.inf)
print(masked)
print(attn_scores)

tensor([[0., 1., 1., 1., 1., 1.],
        [0., 0., 1., 1., 1., 1.],
        [0., 0., 0., 1., 1., 1.],
        [0., 0., 0., 0., 1., 1.],
        [0., 0., 0., 0., 0., 1.],
        [0., 0., 0., 0., 0., 0.]])
tensor([[False,  True,  True,  True,  True,  True],
        [False, False,  True,  True,  True,  True],
        [False, False, False,  True,  True,  True],
        [False, False, False, False,  True,  True],
        [False, False, False, False, False,  True],
        [False, False, False, False, False, False]])
tensor([[0.3111,   -inf,   -inf,   -inf,   -inf,   -inf],
        [0.1655, 0.2602,   -inf,   -inf,   -inf,   -inf],
        [0.1667, 0.2602, 0.2577,   -inf,   -inf,   -inf],
        [0.0510, 0.1080, 0.1064, 0.0643,   -inf,   -inf],
        [0.1415, 0.1875, 0.1863, 0.0987, 0.1121,   -inf],
        [0.0476, 0.1192, 0.1171, 0.0731, 0.0477, 0.0966]],
       grad_fn=<MaskedFillBackward0>)
tensor([[0.3111, 0.3479, 0.3471, 0.1714, 0.2350, 0.1928],
        [0.1655, 0.2602, 0.2576, 0.14

In [113]:
#下一步就是使用softmax做归一化
attn_weights = torch.softmax(masked / keys.shape[-1]**0.5,dim=1)
print(attn_weights)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.4833, 0.5167, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3190, 0.3408, 0.3402, 0.0000, 0.0000, 0.0000],
        [0.2445, 0.2545, 0.2542, 0.2468, 0.0000, 0.0000],
        [0.1994, 0.2060, 0.2058, 0.1935, 0.1953, 0.0000],
        [0.1624, 0.1709, 0.1706, 0.1654, 0.1625, 0.1682]],
       grad_fn=<SoftmaxBackward0>)


##### 5.2 Masking additional attention weights with dropout

深度学习中的 Dropout 是一种技术，在训练过程中随机选择隐藏层单元并将其忽略，实际上是将它们 “丢弃”。这种方法通过确保模型不会过度依赖于任何特定的隐藏层单元集合来帮助防止过拟合。需要强调的是，Dropout 仅在训练期间使用，之后会被禁用。

在 Transformer 架构中，包括像 GPT 这样的模型，注意力机制中的 Dropout 通常在两个特定时间应用：在计算注意力权重之后，或者在将注意力权重应用于值向量之后。在这里，我们将在计算注意力权重之后应用 Dropout 掩码，如图 3.22 所示，因为这在实践中是更常见的变体。

Dropout的原理与作用

随机忽略隐藏层单元：Dropout 在训练过程中，会以一定的概率随机地将部分隐藏层单元的输出置为零。这意味着在每次训练迭代中，模型实际上是在使用一个不同的、简化的子网络进行训练。例如，假设一个神经网络有 100 个隐藏层单元，设置 Dropout 概率为 0.5，那么在每次训练时，平均会有 50 个单元被随机 “丢弃”，模型就像是在使用一个只有部分单元起作用的较小网络进行训练。

防止过拟合：通过这种方式，Dropout 可以有效地防止过拟合。如果模型没有 Dropout，它可能会过度依赖于某些特定的隐藏层单元，从而对训练数据过度拟合，导致在新的、未见过的数据上表现不佳。而有了 Dropout，模型被迫学习更加鲁棒的特征表示，不依赖于特定的单元组合，从而提高了模型的泛化能力。

仅在训练时使用：Dropout 只在训练阶段起作用。在测试或推理阶段，所有的隐藏层单元都会被使用，以充分发挥模型的性能。这是因为在测试阶段，我们希望得到模型的最准确预测，而不是通过随机丢弃单元来增加不确定性。

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-22.png)

如上图所示，我们使用因果注意力掩码（左上角）和额外的dropout 掩码（右上角）来将额外部分注意力权重归置为0，降低训练过程中过拟合。

接下来我们来看下pytorch如何实现dropout的，我们先看下dropout的比率50%，意味着将有一半的权重将会丢弃。实际上GPT模型一般使用0.1或者0.2的比率。

In [160]:
torch.manual_seed(123)
dropout = torch.nn.Dropout(0.5)
example = torch.ones(6,6)
print(example)
print(dropout(example))

tensor([[1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.]])
tensor([[2., 2., 0., 2., 2., 0.],
        [0., 0., 0., 2., 0., 2.],
        [2., 2., 2., 2., 0., 2.],
        [0., 2., 2., 0., 0., 2.],
        [0., 2., 0., 2., 0., 2.],
        [0., 2., 2., 2., 2., 0.]])


由于有一半的元素被设为零，有效元素的数量减少了。为了补偿这种减少，需要对剩下的元素进行缩放。这里提到的缩放因子是 1/(1-0.5)。

从数学角度理解，假设原始注意力权重矩阵某一行的元素之和代表该行的 “总影响力”，在应用 Dropout 后，若不进行缩放，这一 “总影响力” 会因为部分元素变为零而降低。而缩放剩余元素可以使得该行元素的 “总影响力” 在期望上保持不变。

以一个简单的情况为例，假设原始注意力权重矩阵的某一行元素为[a,b,c,d] ，其和为S=a+b+c+d ，经过 Dropout 后，假设只剩下 [a,c]（b和d被设为零），如果不缩放，这两个元素的和为a+c，明显小于原始的S。但如果将剩下的元素缩放 2 倍，变为 ，其和为 2a+2c=2(a+c)，在期望上更接近原始的 “总影响力S”（考虑到 Dropout 是随机的，多次操作的平均效果）

In [117]:
torch.manual_seed(123)
print(dropout(attn_weights))

tensor([[2.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.6380, 0.6816, 0.6804, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.5090, 0.5085, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.4120, 0.0000, 0.3869, 0.0000, 0.0000],
        [0.0000, 0.3418, 0.3413, 0.3308, 0.3249, 0.0000]],
       grad_fn=<MulBackward0>)


##### 5.3 Implementing a compact causal attention class

这里我们将结合上一节和这节讲的内容实现一个更为实用的python类，并作为后面多头自注意力机制的类基础，并在这里支持批量input数据。像在之前讲的dataloader那样。

In [119]:
# 为了模拟批量输入，我们简单复制inputs成两份
batch = torch.stack((inputs,inputs),dim=0)
print(inputs.shape)
print(batch.shape)

torch.Size([6, 3])
torch.Size([2, 6, 3])


In [127]:
class CausalAttention(nn.Module):
    def __init__(self,d_in,d_out,context_length,dropout,qkv_bias=False):
        super().__init__()
        self.d_out = d_out
        self.W_query = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.W_key = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.W_value = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.dropout = nn.Dropout(dropout)
        self.register_buffer(
            'mask',
            torch.triu(torch.ones(context_length,context_length),
                       diagonal=1)
        )
    def forward(self,x):
        b,num_tokens,d_in = x.shape
        keys = self.W_key(x)
        queries = self.W_query(x)
        values = self.W_value(x)

        # 这里注意keys的转置，因为第0维是批次
        attn_scores = queries @ keys.transpose(1,2)
        attn_scores.masked_fill_(
            self.mask.bool()[:num_tokens,:num_tokens],-torch.inf
        )
        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5,dim=-1
        )
        attn_weights = self.dropout(attn_weights)
        context_vec = attn_weights @ values
        return context_vec


上面我们有两个新的知识点：

self.register_buffer()，register_buffer()有两个作用：1.模型里有需要训练的参数tensor，还有不需要训练更新的tensor变量，不需要更新的变量就可以放在缓冲器里，这样每次在迭代时候，就不用计算这些tensor的梯度和更新了。同时方便显示哪些变量是需要训练的参数2.如果是需要训练的参数，pytorch会自动在设备之间进行迁移数据，而不需要训练更新的变量则需要自行迁移了，比如从CPU到GPU的过程。register_buffer来维护这个迁移过程。



In [128]:
torch.manual_seed(123)
context_length = batch.shape[1]
ca = CausalAttention(d_in,d_out,context_length,0.0)
context_vecs = ca(batch)
print("context_vecs.shape",context_vecs.shape)

context_vecs.shape torch.Size([2, 6, 2])


![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-23.png)

#### 6.Extending single-head attention to multi-head attention

最后一步是扩展因果注意力到多头注意力(multi-head attention).

“多头”其实是指多个因果注意力类实例并行执行，互相独立。

我们先从最简单的做起，就是创建一个多头注意力模块通过叠加多个CausalAttention模块。然后再实现一个更加高效的多头注意力模块。

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-24.png)

从上图可以看出，我们使用两个不同的权重参数矩阵，分别计算出context vector，最后将两个context vector合并在一起即可。

In [129]:
class MultiHeadAttentionWrapper(nn.Module):
    def __init__(self,d_in,d_out,context_length,dropout,num_heads,qkv_bias=False):
        super().__init__()
        self.heads = nn.ModuleList(
            [CausalAttention(
                d_in,d_out,context_length,dropout,qkv_bias
            ) for _ in range(num_heads)]
        )
    def forward(self,x):
        return torch.cat([head(x) for head in self.heads],dim=-1)

![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-25.png)

假设使用两个头的话，那么context vector的维度就是d_out * head_number=2*2=4

In [133]:
torch.manual_seed(123)
num_heads = 2
context_length = batch.shape[1]
mhaw = MultiHeadAttentionWrapper(d_in,d_out,context_length,0.0,num_heads)
context_vecs = mhaw(batch)
print("context_vecs.shape",context_vecs.shape)
print(context_vecs)

context_vecs.shape torch.Size([2, 6, 4])
tensor([[[-0.4519,  0.2216,  0.4772,  0.1063],
         [-0.5874,  0.0058,  0.5891,  0.3257],
         [-0.6300, -0.0632,  0.6202,  0.3860],
         [-0.5675, -0.0843,  0.5478,  0.3589],
         [-0.5526, -0.0981,  0.5321,  0.3428],
         [-0.5299, -0.1081,  0.5077,  0.3493]],

        [[-0.4519,  0.2216,  0.4772,  0.1063],
         [-0.5874,  0.0058,  0.5891,  0.3257],
         [-0.6300, -0.0632,  0.6202,  0.3860],
         [-0.5675, -0.0843,  0.5478,  0.3589],
         [-0.5526, -0.0981,  0.5321,  0.3428],
         [-0.5299, -0.1081,  0.5077,  0.3493]]], grad_fn=<CatBackward0>)


这里多头是串行计算，效率比较低下，接下来我们实现比较高效的方法

在此之前我们先学习view和transpose的方法

In [146]:
t = torch.rand(2,3,4)
print(t)

tensor([[[0.2251, 0.3111, 0.1955, 0.9153],
         [0.7751, 0.6749, 0.1166, 0.8858],
         [0.6568, 0.8459, 0.3033, 0.6060]],

        [[0.9882, 0.8363, 0.9010, 0.3950],
         [0.8809, 0.1084, 0.5432, 0.2185],
         [0.3834, 0.3720, 0.5374, 0.9551]]])


In [147]:
t = t.view(2,3,2,2)
print(t)

tensor([[[[0.2251, 0.3111],
          [0.1955, 0.9153]],

         [[0.7751, 0.6749],
          [0.1166, 0.8858]],

         [[0.6568, 0.8459],
          [0.3033, 0.6060]]],


        [[[0.9882, 0.8363],
          [0.9010, 0.3950]],

         [[0.8809, 0.1084],
          [0.5432, 0.2185]],

         [[0.3834, 0.3720],
          [0.5374, 0.9551]]]])


In [148]:
print(t.transpose(1,2))

tensor([[[[0.2251, 0.3111],
          [0.7751, 0.6749],
          [0.6568, 0.8459]],

         [[0.1955, 0.9153],
          [0.1166, 0.8858],
          [0.3033, 0.6060]]],


        [[[0.9882, 0.8363],
          [0.8809, 0.1084],
          [0.3834, 0.3720]],

         [[0.9010, 0.3950],
          [0.5432, 0.2185],
          [0.5374, 0.9551]]]])


在日常编程中，一提到提高任务处理速度，大家立即会想到使用多线程或者多进程的方式，但多线程或者多进程并不是唯一的方式，还有一种办法就是使用更高效的算法压榨单线程的处理性能。

In [157]:
class MultiHeadAttention(nn.Module):
    def __init__(self,d_in,d_out,context_length,dropout,num_heads,qkv_bias=False):
        super().__init__()
        self.d_out = d_out
        self.num_heads = num_heads
        self.head_dim = d_out // num_heads
        self.W_query = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.W_key = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.W_value = nn.Linear(d_in,d_out,bias=qkv_bias)
        self.dropout = nn.Dropout(dropout)
        self.out_proj = nn.Linear(d_out,d_out)
        self.register_buffer(
            'mask',
            torch.triu(torch.ones(context_length,context_length),
                       diagonal=1)
        )
    def forward(self,x):
        b,num_tokens,d_in = x.shape
        keys = self.W_key(x)
        queries = self.W_query(x)
        values = self.W_value(x)

        # 将keys,queries,values转化为多头的模式
        keys = keys.view(b,num_tokens,self.num_heads,self.head_dim)
        queries = queries.view(b,num_tokens,self.num_heads,self.head_dim)
        values = values.view(b,num_tokens,self.num_heads,self.head_dim)

        keys = keys.transpose(1,2)
        queries = queries.transpose(1,2)
        values = values.transpose(1,2)

        # 这里注意keys的转置，因为第0维是批次
        attn_scores = queries @ keys.transpose(2,3)
        attn_scores.masked_fill_(
            self.mask.bool()[:num_tokens,:num_tokens],-torch.inf
        )
        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5,dim=-1
        )
        attn_weights = self.dropout(attn_weights)
        context_vec = (attn_weights @ values).transpose(1,2)
        # contiguous方法是将张量的数据在内存中是连续的，不然view方法就会报错
        context_vec = context_vec.contiguous().view(
            b,num_tokens,self.d_out
        )
        context_vec = self.out_proj(context_vec)
        return context_vec


![](https://drek4537l1klr.cloudfront.net/raschka/Figures/3-26.png)

我们来看看批量矩阵相乘和单次矩阵相乘的结果

In [149]:
a = torch.tensor([[[[0.2745, 0.6584, 0.2775, 0.8573],
                    [0.8993, 0.0390, 0.9268, 0.7388],
                    [0.7179, 0.7058, 0.9156, 0.4340]],

                   [[0.0772, 0.3565, 0.1479, 0.5331],
                    [0.4066, 0.2318, 0.4545, 0.9737],
                    [0.4606, 0.5159, 0.4220, 0.5786]]]])
print(a.shape)

torch.Size([1, 2, 3, 4])


In [150]:
print(a @ a.transpose(2,3))

tensor([[[[1.3208, 1.1631, 1.2879],
          [1.1631, 2.2150, 1.8424],
          [1.2879, 1.8424, 2.0402]],

         [[0.4391, 0.7003, 0.5903],
          [0.7003, 1.3737, 1.0620],
          [0.5903, 1.0620, 0.9912]]]])


In [151]:
first_head = a[0,0,:,:]
# 对于高于二维的张量，.T操作默认只交换最后两个维度
first_res = first_head @ first_head.T
print("Fist head:\n",first_res)

second_head = a[0,1,:,:]
# 对于高于二维的张量，.T操作默认只交换最后两个维度
second_res = second_head @ second_head.T
print("Second head:\n",second_res)

Fist head:
 tensor([[1.3208, 1.1631, 1.2879],
        [1.1631, 2.2150, 1.8424],
        [1.2879, 1.8424, 2.0402]])
Second head:
 tensor([[0.4391, 0.7003, 0.5903],
        [0.7003, 1.3737, 1.0620],
        [0.5903, 1.0620, 0.9912]])


In [158]:
torch.manual_seed(123)
batch_size, context_length,d_in = batch.shape
d_out = 2
mha = MultiHeadAttention(d_in,d_out,context_length,0.0,num_heads=2)
context_vecs = mha(batch)
print(context_vecs)
print("context_vecs.shape:",context_vecs.shape)


tensor([[[0.3190, 0.4858],
         [0.2943, 0.3897],
         [0.2856, 0.3593],
         [0.2693, 0.3873],
         [0.2639, 0.3928],
         [0.2575, 0.4028]],

        [[0.3190, 0.4858],
         [0.2943, 0.3897],
         [0.2856, 0.3593],
         [0.2693, 0.3873],
         [0.2639, 0.3928],
         [0.2575, 0.4028]]], grad_fn=<ViewBackward0>)
context_vecs.shape: torch.Size([2, 6, 2])


While using the multi-head attention based on scaled-dot product attention remains the most common variant of self-attention in practice, authors have found that it’s possible to also achieve good performance without the value weight matrix and projection layer:

“Simplifying Transformer Blocks” (2023) by He and Hofmann, https://arxiv.org/abs/2311.01906